packages/frontend/src/redux/reducers/sign/index.js

```diff
@@ -25,43 +46,73 @@ const sign = handleActions({
                 .reduce((totalAmount, amount) => totalAmount.add(new BN(amount)), new BN(0)).toString(),
             fees: {
                 transactionFees: '', // TODO: Calculate total fees
-                gasLimit: allActions
-                    .filter(a => Object.keys(a)[0] === 'functionCall')
-                    .map(a => a.functionCall.gas)
-                    .reduce((totalGas, gas) => totalGas.add(gas), new BN(0)).toString(),
+                gasLimit: calculateGasLimit(allActions),
                 gasPrice: '' // TODO: Where to get gas price?
             },
             sensitiveActionsCounter: allActions
                 .filter(a => ['deployContract', 'stake', 'deleteAccount'].indexOf(Object.keys(a)[0]) > -1)
                 .length
         };
     },
-    [signAndSendTransactions]: (state, { error, payload, ready }) => {
+    [handleSignTransactions.pending]: (state) => ({
+        ...state,
+        status: SIGN_STATUS.IN_PROGRESS
+    }),
+    [handleSignTransactions.fulfilled]: (state, { payload }) => ({
+        ...state,
+        status: SIGN_STATUS.SUCCESS,
+        transactionHashes: payload
+    }),
+    [handleSignTransactions.rejected]: (state, { error }) => {
+        const retryTxDirection = error.message.includes('Exceeded the prepaid gas')
+            ? RETRY_TX.INCREASE
+            : error.message.includes('TotalPrepaidGasExceeded')
+                ? RETRY_TX.DECREASE
+                : undefined;
+        
+        const tryRetryTx = retryTxDirection && !state.transactions.some((t) => 
+            t.actions && t.actions.some((a) => a.functionCall && a.functionCall.gas && (
+                (
+                    state.retryTxDirection === RETRY_TX.INCREASE 
+                    && a.functionCall.gas.gt(new BN(RETRY_TX.GAS.MAX))
+                ) || (
+                    state.retryTxDirection === RETRY_TX.DECREASE 
+                    && a.functionCall.gas.lte(new BN(RETRY_TX.GAS.MIN))
+                )
+            ))
+        );
 
-        if (!ready) {
-            return {
-                ...state,
-                status: 'needs-confirmation'
-            };
+        if (retryTxDirection && !tryRetryTx) {
+            Mixpanel.track('SIGN - RETRY - retry limit exceeded');
         }
 
-        if (error) {
-            return {
-                ...state,
-                status: 'error',
-                error: payload
-            };
+        const transactions = cloneDeep(state.transactions);
```

Comment:
If we're not mutating `state.transactions`, we **shouldn't** clone it and replace the existing `state.transactions` with a clone, we should just leave `state.transactions` as-is.

Reply:
Right. Done.